### PR TITLE
Add feature toggle and placeholder for reentry form flow

### PR DESCRIPTION
**Pull request by shanear**
Because adding the RAMP re-entry form is going to be a big
feature change implemented over many PRs and revisions,
were gonna slide it behind a feature flag.

## Test instructions
- go to the intake start page, nothing should have changed
- enable the `:intake_reentry_form` feature toggle
- go to the intake start page and see the form select page (it wont do anything for now)

connects #3956

### CODE REVIEW

**Comment by nickheiner-usds** [comment]
> - go to the intake start page and see the form select page (it wont do anything for now)

I assume it means that this is the intended behavior for clicking "Continue to search"?

image

File: app/views/intakes/index.html.erb [hunk 1]
```diff
@@ -5,6 +5,9 @@
     dropdownUrls: dropdown_urls,
     page: "Intake",
     feedbackUrl: ENV["CASEFLOW_FEEDBACK_URL"],
-    buildDate: build_date
+    buildDate: build_date,
+    featureToggles: {
```
Comment (nickheiner-usds) [inline app/views/intakes/index.html.erb]: Frontend feature toggles NICE

File: client/app/intake/pages/selectForm.jsx [hunk 2]
```diff
@@ -0,0 +1,66 @@
...
+    </Button>;
+}
+
+const SelectFormButtonConnected = connect(
+  ({ formSelection }) => ({ formSelection }),
+)(SelectFormButton);
+
+export class SelectFormButtons extends React.PureComponent {
+  render = () =>
+    <div>
```
Comment (nickheiner-usds) [inline client/app/intake/pages/selectForm.jsx]: This `<div>` is probably unnecessary.

File: client/app/intake/redux/actions.js [hunk 3]
```diff
@@ -76,6 +76,18 @@ export const setOptionSelected = (optionSelected) => ({
   }
 });
 
+export const setFormSelection = (formSelection) => ({
+  type: ACTIONS.SET_FORM_SELECTION,
+  payload: {
+    formSelection
+  },
+  meta: {
```
Comment (nickheiner-usds) [inline client/app/intake/redux/actions.js]: MMMHMMMM ANALYTICS :1st_place_medal:

File: spec/feature/intake_spec.rb [hunk 4]
```diff
@@ -310,6 +310,22 @@
...
+      before { FeatureToggle.enable!(:intake_reentry_form) }
+      after { FeatureToggle.disable!(:intake_reentry_form) }
+
+      scenario "flow starts with form selection" do
+        visit "/intake"
+
+        within_fieldset("Which form are you processing?") do
+          find("label", text: "21-4138 RAMP Selection Form").click
+        end
+        safe_click ".cf-submit.usa-button"
```
Comment (nickheiner-usds) [inline spec/feature/intake_spec.rb]: 🏩

File: client/app/intake/IntakeFrame.jsx [hunk 5]
```diff
@@ -74,28 +76,67 @@ class IntakeFrame extends React.PureComponent {
                   lowerMargin
                 />
               }
-              <PageRoute
```
Comment (nickheiner-usds) [inline client/app/intake/IntakeFrame.jsx]: This can be simplified:

```diff
diff --git a/client/app/intake/IntakeFrame.jsx b/client/app/intake/IntakeFrame.jsx
index 56f3ff0..bbd4d0c 100644
--- a/client/app/intake/IntakeFrame.jsx
+++ b/client/app/intake/IntakeFrame.jsx
@@ -41,6 +41,25 @@ class IntakeFrame extends React.PureComponent {
       cancelButton = <Button linkStyling onClick={this.props.toggleCancelModal} id="close-modal">Close</Button>;
     }
 
+    const firstPages = this.props.featureToggles.reentryForm ?
+      <div>
+        <PageRoute
+        exact
+        path={PAGE_PATHS.BEGIN}
+        title="Select Form | Caseflow Intake"
+        component={SelectFormPage} />
+      <PageRoute
+        exact
+        path={PAGE_PATHS.SEARCH}
+        title="Search | Caseflow Intake"
+        component={SearchPage} />
+      </div> :
+      <PageRoute
+        exact
+        path={PAGE_PATHS.BEGIN}
+        title="Begin Intake | Caseflow Intake"
+        component={BeginPage} />;
+
     return <Router basename="/intake" {...this.props.routerTestProps}>
       <div>
         { this.props.cancelModalVisible &&
@@ -76,18 +95,8 @@ class IntakeFrame extends React.PureComponent {
                   lowerMargin
                 />
               }
-              { this.props.featureToggles.reentryForm &&
                 <div>
-                  <PageRoute
-                    exact
-                    path={PAGE_PATHS.BEGIN}
-                    title="Select Form | Caseflow Intake"
-                    component={SelectFormPage} />
-                  <PageRoute
-                    exact
-                    path={PAGE_PATHS.SEARCH}
-                    title="Search | Caseflow Intake"
-                    component={SearchPage} />
+                  {firstPages}
                   <PageRoute
                     exact
                     path={PAGE_PATHS.REVIEW}
@@ -104,31 +113,6 @@ class IntakeFrame extends React.PureComponent {
                     title="Confirmation | Caseflow Intake"
                     component={CompletedPage} />
                 </div>
-              }
-              { !this.props.featureToggles.reentryForm &&
-                <div>
-                  <PageRoute
-                    exact
-                    path={PAGE_PATHS.BEGIN}
-                    title="Begin Intake | Caseflow Intake"
-                    component={BeginPage} />
-                  <PageRoute
-                    exact
-                    path={PAGE_PATHS.REVIEW}
-                    title="Review Election | Caseflow Intake"
-                    component={ReviewPage} />
-                  <PageRoute
-                    exact
-                    path={PAGE_PATHS.FINISH}
-                    title="Finish | Caseflow Intake"
-                    component={FinishPage} />
-                  <PageRoute
-                    exact
-                    path={PAGE_PATHS.COMPLETED}
-                    title="Completed | Caseflow Intake"
-                    component={CompletedPage} />
-                </div>
-              }
             </PrimaryAppContent>
             <AppSegment className="cf-workflow-button-wrapper">
               { this.props.featureToggles.reentryForm &&

```

File: client/app/intake/IntakeFrame.jsx [hunk 6]
```diff
@@ -74,28 +76,67 @@ class IntakeFrame extends React.PureComponent {
...
+                  <PageRoute
+                    exact
+                    path={PAGE_PATHS.COMPLETED}
+                    title="Completed | Caseflow Intake"
+                    component={CompletedPage} />
+                </div>
+              }
             </PrimaryAppContent>
             <AppSegment className="cf-workflow-button-wrapper">
+              { this.props.featureToggles.reentryForm &&
```
Comment (nickheiner-usds) [inline client/app/intake/IntakeFrame.jsx]: 👍

File: client/app/intake/pages/selectForm.jsx [hunk 7]
```diff
@@ -0,0 +1,66 @@
...
+      onClick={this.handleClick}
+      legacyStyling={false}
+      disabled={!this.props.formSelection}
+    >
+      Continue to search
+    </Button>;
+}
+
+const SelectFormButtonConnected = connect(
+  ({ formSelection }) => ({ formSelection }),
```
Comment (nickheiner-usds) [inline client/app/intake/pages/selectForm.jsx]: You can use this same syntax when `connect`ing `SelectForm` above.

File: client/app/intake/pages/selectForm.jsx [hunk 8]
```diff
@@ -0,0 +1,66 @@
...
+)(SelectFormButton);
+
+export class SelectFormButtons extends React.PureComponent {
```
Comment (nickheiner-usds) [inline client/app/intake/pages/selectForm.jsx]: In fact, this whole component is unnecessary. We can just `export` `SelectFormButtonConnected` directly.

```diff
diff --git a/client/app/intake/pages/selectForm.jsx b/client/app/intake/pages/selectForm.jsx
index a324b72..f68d4c1 100644
--- a/client/app/intake/pages/selectForm.jsx
+++ b/client/app/intake/pages/selectForm.jsx
@@ -38,7 +38,7 @@ export default connect(
   }, dispatch)
 )(SelectForm);
 
-class SelectFormButton extends React.PureComponent {
+class UnconnectedSelectFormButton extends React.PureComponent {
   handleClick = () => {
     this.props.history.push('/search');
   }
@@ -54,13 +54,6 @@ class SelectFormButton extends React.PureComponent {
     </Button>;
 }
 
-const SelectFormButtonConnected = connect(
+export const SelectFormButton = connect(
   ({ formSelection }) => ({ formSelection }),
-)(SelectFormButton);
-
-export class SelectFormButtons extends React.PureComponent {
-  render = () =>
-    <div>
-      <SelectFormButtonConnected history={this.props.history} />
-    </div>
-}
+)(UnconnectedSelectFormButton);

```

File: spec/feature/intake_spec.rb [hunk 9]
```diff
@@ -310,6 +310,22 @@
       visit "/intake/finish"
       expect(page).to have_content("Welcome to Caseflow Intake!")
     end
+
+    context "when ramp reentry form is enabled", focus: true do
+      before { FeatureToggle.enable!(:intake_reentry_form) }
```
Comment (nickheiner-usds) [inline spec/feature/intake_spec.rb]: 👍

File: spec/feature/intake_spec.rb [hunk 10]
```diff
@@ -310,6 +310,22 @@
       visit "/intake/finish"
       expect(page).to have_content("Welcome to Caseflow Intake!")
     end
+
+    context "when ramp reentry form is enabled", focus: true do
```
Comment (nickheiner-usds) [inline spec/feature/intake_spec.rb]: `focus: true` should be removed before merging, of course. :smile:

File: client/app/intake/IntakeFrame.jsx (hunk 5 above)
Comment (shanear) [reply client/app/intake/IntakeFrame.jsx]: I'm going to end up having to split these back out in a future PR, so im just going to keep them separate.

**Comment by nickheiner-usds** [comment]
Here's my suggestion for switching on form type throughout the app in a minimally-wordy way:

```diff
diff --git a/client/app/intake/IntakeFrame.jsx b/client/app/intake/IntakeFrame.jsx
index 56f3ff0..b8208f2 100644
--- a/client/app/intake/IntakeFrame.jsx
+++ b/client/app/intake/IntakeFrame.jsx
@@ -21,6 +21,29 @@ import CompletedPage, { CompletedNextButton } from './pages/completed';
 import { PAGE_PATHS, REQUEST_STATE } from './constants';
 import { toggleCancelModal, submitCancel } from './redux/actions';
 
+class SwitchOnForm extends React.PureComponent {
+  render = () => {
+    const { formSelection, formComponentMapping, componentForNoFormSelected } = this.props;
+
+    if (!formSelection) {
+      return componentForNoFormSelected;
+    }
+
+    const child = formComponentMapping[formSelection];
+
+    // eslint-disable-next-line no-undefined
+    if (child === undefined) {
+      throw new Error(`SwitchOnForm does not have a mapping for current form selection: '${formSelection}'`);
+    }
+
+    return child;
+  }
+}
+
+const ConnectedSwitchOnForm = connect(
+  ({ formSelection }) => ({ formSelection })
+)(SwitchOnForm);
+
 class IntakeFrame extends React.PureComponent {
   handleSubmitCancel = () => (
     this.props.submitCancel(this.props.rampElection)
@@ -63,6 +86,13 @@ class IntakeFrame extends React.PureComponent {
           topMessage={topMessage}
           defaultUrl="/">
           <AppFrame>
+            <ConnectedSwitchOnForm
+              formComponentMapping={{
+                ramp_election: <p>Component: Ramp Election</p>,
+                ramp_reentry: <p>Component: Ramp Reentry</p>
+              }}
+              componentForNoFormSelected={<p>Component: no form selected</p>}
+            />
             <IntakeProgressBar />
             <PrimaryAppContent>
               { this.props.requestStatus.cancelIntake === REQUEST_STATE.FAILED &&

```

**Pull request merged**